analyzers/ruby/ruby.go

```diff
@@ -127,16 +127,26 @@ func (a *Analyzer) IsBuilt() (bool, error) {
 }
 
 func (a *Analyzer) Analyze() (graph.Deps, error) {
+	strategy := "list"
+
+	if a.Options.Strategy != "" {
+		strategy = a.Options.Strategy
+	}
+
+	return a.analyzeHelper(strategy)
+}
+
+func (a *Analyzer) analyzeHelper(strategy string) (graph.Deps, error) {
 	lockfilePath := filepath.Join(a.Module.Dir, "Gemfile.lock")
 	if a.Options.LockfilePath != "" {
 		lockfilePath = a.Options.LockfilePath
 	}
 
-	switch a.Options.Strategy {
+	switch strategy {
 	case "list":
 		gems, err := a.Bundler.List()
 		if err != nil {
-			return graph.Deps{}, err
+			return a.analyzeHelper("lockfile")
```

Comment:
The potential for accidental infinite recursion is a bit scary here. Given that this function is a gigantic switch anyway, I think it makes sense to break out these strategies into separate functions.